source/code/plugins/agent_maintenance_script.rb

```diff
@@ -171,32 +183,105 @@ def apply_certificate_update_endpoint(server_resp, check_for_renew_request = tru
 
     # Update the DSC_ENDPOINT variable in omsadmin.conf from the server XML
     def apply_dsc_endpoint(server_resp)
-      dsc_endpoint = ""
+      @pending_DSC_ENDPOINT = ""
 
       # Extract the DSC endpoint from the server response
       dsc_conf_regex = /<DscConfiguration.*<Endpoint>(?<endpoint>.*)<\/Endpoint>.*DscConfiguration>/
       dsc_conf_regex.match(server_resp) { |match|
-        dsc_endpoint = match["endpoint"]
+        @pending_DSC_ENDPOINT = match["endpoint"]
         # Insert escape characters before open and closed parentheses
-        dsc_endpoint = dsc_endpoint.gsub("(", "\\\\(").gsub(")", "\\\\)")
+        @pending_DSC_ENDPOINT = @pending_DSC_ENDPOINT.gsub("(", "\\\\(").gsub(")", "\\\\)")
       }
 
-      if dsc_endpoint.empty?
+      if @pending_DSC_ENDPOINT.empty?
         log_error("Could not extract the DSC endpoint.")
         return OMS::ERROR_EXTRACTING_ATTRIBUTES
       end
 
       # Update omsadmin.conf with dsc_endpoint variable
       # When apply_dsc_endpoint is called from onboarding, dsc_endpoint will be returned in file
-      update_config("DSC_ENDPOINT", dsc_endpoint)
+      if @current_DSC_ENDPOINT != @pending_DSC_ENDPOINT
+        log_info("DSC endpoint change detected, old: #{@current_DSC_ENDPOINT}, new: #{@pending_DSC_ENDPOINT}. Updating omsadmin.conf ...")
+        update_config("DSC_ENDPOINT", @pending_DSC_ENDPOINT)
+        @current_DSC_ENDPOINT = @pending_DSC_ENDPOINT
+        @pending_metaconfig_generation = true
+      end
+
+      return @current_DSC_ENDPOINT
+    end
+
+    # Generate the DSC MetaConfig file. If the dsc endpoint is updated, DSC will
+    # not directly pick the endpoint from omsadmin.conf (instead, it is picked
+    # from MetaConfig.mof), so we must manually trigger metaconfig (re)generation.
+    def generate_dsc_metaconfig
+      # Avoid calling the generation script if unnecessary
+      return if !@pending_metaconfig_generation
+
+      # If DSC is disabled, we delay MetaConfig generation until the next topology request where it has been re-enabled
+      if OMS::Config.is_dsc_disabled
+        log_info("DSC has been disabled, skipping DSC metaconfig generation")
+        return
+      end
+
+      retries = 0
+      max_retries = 3
+      log_info("Beginning DSC megaconfig generation")
+      while retries < max_retries
+        success = call_oms_metaconfighelper
+        if success
+          @pending_metaconfig_generation = false
+          return
+        else
+          sleep 5
+          retries += 1
+        end
+      end
+
+      OMS::Log.warn_once("Exceeded max attempts to generate DSC metaconfig, will retry after next topology request")
+    end
+
+    # Call OMS_MetaConfigHelper.py to trigger metaconfig generation, return success
+    def call_oms_metaconfighelper
+      # Note that OMS_MetaConfigHelper.py will log to omsconfig.log
+      command = "/opt/microsoft/omsconfig/Scripts/OMS_MetaConfigHelper.py"
```

Comment:
this will have to change once [this](https://github.com/microsoft/PowerShell-DSC-for-Linux/pull/691/files) is taken